src/storage/src/source/generator/counter.rs

```diff
@@ -22,14 +27,50 @@ impl Generator for Counter {
         _seed: Option<u64>,
     ) -> Box<dyn Iterator<Item = (usize, GeneratorMessageType, Row, i64)>> {
         let mut counter = 0;
-        Box::new(iter::repeat_with(move || {
-            counter += 1;
-            (
-                0,
-                GeneratorMessageType::Finalized,
-                Row::pack_slice(&[Datum::Int64(counter)]),
-                1,
-            )
-        }))
+        let max_cardinality = self.max_cardinality;
+        Box::new(
+            iter::repeat_with(move || {
+                let to_retract = match max_cardinality {
+                    Some(max) => {
+                        if max <= counter {
+                            Some(counter - max + 1)
```

Comment:
Ah this line has changed the implied type of counter from i64 to usize. Before we had the nice property that it'd rollover at i64 max, now it panics instead and we have 2 try_into's. It'd be nice to never panic and never call try_into.

What if max_cardinality is an `Option<i64>` instead (contrary to my comment above)? The planner step could do the validation to ensure it's > 0, and then maybe all the math gets very clean?

Oh it'd still have the property that after rollever the max cardinality stops working because of the `max <= counter` check. That could be fixed by having two counters: `counter` and `until_retraction` that would start at `max_cardinality` and either subtract 1 from itself or if it's 0 then stay at 0 and do the retraction.

Does any of this matter because i64 is still a number high enough we'll never ever see it even at the lowest tick interval possible?

Reply:
The lowest tick interval is 1µs, which will take hundreds of thousands of years to overflow an i64. Given that fact, I don't think it makes sense to make the code more complicated to support the overflow case.

Reply:
I think I'm ok not doing all of ^ if it's really annoying as long as we add a comment saying it'll never panic in practice because even at a tick interval of 1ns it'll still take X millennia before it triggers or something.

Hmm what if we ever want to support a tick interval of 0 to simulate a busy source? Maybe it should be safer.